### DISCUSSION TITLE
Can't work out the documentation for adding a PnPCoreAuthentication provider

### DISCUSSION THREAD

**Discussion by juliemturner** [pnp/pnpcore#717] [Q&A] [answered]
I'm working on creating an Azure Function v3 with version 1.5 of PnP.Core.Auth. I would like to create a certification authentication provider.  The example you have is for v4 so I'm having to try and figure out how to make this work in v3 and I'm getting an error when trying to create an instance of the auth provider before passing it to the context. The first line below throws an "System.NullReferenceException: 'Object reference not set to an instance of an object.'" (see stack trace below) although I can see that my auth provider is actually registered in the factory->Non-Public Members->Options->Credentials->Configurations

``` C#
var certAuthProvider = _pnpAuthProviderFactory.Create("CertAuth");
using (var pnpContext = _pnpContextFactory.Create(new Uri(webHook.WebUrl), certAuthProvider))
{
 ...
}
```

I registered the auth provider like this in the Startup.cs file:

``` C#
        public override void Configure(IFunctionsHostBuilder builder)
        {
            builder.Services.AddPnPCore();
            builder.Services.AddPnPCoreAuthentication(options =>
            {
                // Configure certificate based auth
                options.Credentials.Configurations.Add("CertAuth", new PnPCoreAuthenticationCredentialConfigurationOptions
                {
                    ClientId = Environment.GetEnvironmentVariable("ClientId"),
                    TenantId = Environment.GetEnvironmentVariable("TenantId"),
                    X509Certificate = new PnPCoreAuthenticationX509CertificateOptions
                    {
                        Certificate = LoadCertificate(),
                    }
                });

            });
        }
```

So, i guess my question is, can I not do this in v3?

Stack Trace:
   at DryIoc.Container.ResolveAndCacheDefaultFactoryDelegate(Type serviceType, IfUnresolved ifUnresolved) in D:\a\1\s\src\WebJobs.Script.WebHost\DependencyInjection\DryIoc\Container.cs:line 225
   at DryIoc.Container.DryIoc.IResolver.Resolve(Type serviceType, IfUnresolved ifUnresolved) in D:\a\1\s\src\WebJobs.Script.WebHost\DependencyInjection\DryIoc\Container.cs:line 196
   at Microsoft.Azure.WebJobs.Script.WebHost.DependencyInjection.JobHostServiceProvider.GetService(Type serviceType, IfUnresolved ifUnresolved) in D:\a\1\s\src\WebJobs.Script.WebHost\DependencyInjection\JobHostServiceProvider.cs:line 99
   at Microsoft.Azure.WebJobs.Script.WebHost.DependencyInjection.JobHostServiceProvider.GetService(Type serviceType) in D:\a\1\s\src\WebJobs.Script.WebHost\DependencyInjection\JobHostServiceProvider.cs:line 77
   at PnP.Core.Auth.Services.AuthenticationProviderFactory.Create(String name)
   at LibrarySubscritption.ProcessItemNotification.ProcessNotification(NotificationModel notification) in C:\code\Public-Samples\Webhooks\WebhooksDemo\common\ProcessItemNotification.cs:line 40

**Comment by jansenbe** [comment, marked as answer]
Hi @juliemturner ,

This sample https://pnp.github.io/pnpcore/demos/Demo.AzureFunction/README.html uses Azure V3 functions with certificate based auth via the `X509CertificateAuthenticationProvider` provider but your code is almost OK as well, the issue your seeing comes from the fact that you've not connected a site configuration with the auth provider. Can you try something along these lines and see if that works for you:

```csharp
// Add and configure PnP Core SDK
builder.Services.AddPnPCore(options =>
{
    // Add the base site url
    options.Sites.Add("Default", new PnPCoreSiteOptions
    {
        SiteUrl = Environment.GetEnvironmentVariable("SiteUrl")
    });
});

builder.Services.AddPnPCoreAuthentication(options =>
{
    // Load the certificate to use
    X509Certificate2 cert = LoadCertificate(azureFunctionSettings);

    // Configure certificate based auth
    options.Credentials.Configurations.Add("CertAuth", new PnPCoreAuthenticationCredentialConfigurationOptions
    {
        ClientId = azureFunctionSettings.ClientId,
        TenantId = azureFunctionSettings.TenantId,
        X509Certificate = new PnPCoreAuthenticationX509CertificateOptions
        {
            Certificate = LoadCertificate(azureFunctionSettings),
        }
    });

    // Connect this auth method to the configured site
    options.Sites.Add("Default", new PnPCoreAuthenticationSiteOptions
    {
        AuthenticationProviderName = "CertAuth",
    });
});
```

**Reply by juliemturner** [reply]
Yes, I saw the V3 sample as well that's what I'm saying I'm trying to merge them together... the problem I have with adding the site to the config as you've shown is that the site is dynamic so I would never use it.  The solution I shared with you before just lets me create an auth provider and use it so I assumed I could still do that.  I suppose I could just fake out the root site collection URL with the tenant name property but it felt like a waste because I never use it.

**Reply by juliemturner** [reply]
So I made the changes you suggested, but there is something wrong still with my startup code as it's throwing this error before it even gets to the point of trying to establish context.

![image](https://user-images.githubusercontent.com/7570936/149159116-e7c536a2-50fc-4239-927e-a5ea20d375e4.png)

This error is happening right after the builder tries to add the PnPCore service.  I've validated the siteUrl I built is correct and points to the root site collection in the tenant.

```C#
public override void Configure(IFunctionsHostBuilder builder)
{
    builder.Services.AddPnPCore(options =>
    {
        var siteUrl = $"https://{Environment.GetEnvironmentVariable("Tenant")}.sharepoint.com";
        options.Sites.Add("Default", new PnPCoreSiteOptions
        {
            SiteUrl = siteUrl
        });
    });
    builder.Services.AddPnPCoreAuthentication(options =>
    {
        // Configure certificate based auth
        options.Credentials.Configurations.Add("CertAuth", new PnPCoreAuthenticationCredentialConfigurationOptions
        {
            ClientId = Environment.GetEnvironmentVariable("ClientId"),
            TenantId = Environment.GetEnvironmentVariable("TenantId"),
            X509Certificate = new PnPCoreAuthenticationX509CertificateOptions
            {
                Certificate = LoadCertificate(),
            }
        });
        // Connect this auth method to the configured site
        options.Sites.Add("Default", new PnPCoreAuthenticationSiteOptions
        {
            AuthenticationProviderName = "CertAuth",
        });
    });
}
```
The stack trace is:
   at DryIoc.Container.ResolveAndCacheDefaultFactoryDelegate(Type serviceType, IfUnresolved ifUnresolved) in D:\a\1\s\src\WebJobs.Script.WebHost\DependencyInjection\DryIoc\Container.cs:line 225
   at DryIoc.Container.DryIoc.IResolver.Resolve(Type serviceType, IfUnresolved ifUnresolved) in D:\a\1\s\src\WebJobs.Script.WebHost\DependencyInjection\DryIoc\Container.cs:line 196
   at Microsoft.Azure.WebJobs.Script.WebHost.DependencyInjection.JobHostServiceProvider.GetService(Type serviceType, IfUnresolved ifUnresolved) in D:\a\1\s\src\WebJobs.Script.WebHost\DependencyInjection\JobHostServiceProvider.cs:line 99
   at Microsoft.Azure.WebJobs.Script.WebHost.DependencyInjection.JobHostServiceProvider.GetService(Type serviceType) in D:\a\1\s\src\WebJobs.Script.WebHost\DependencyInjection\JobHostServiceProvider.cs:line 77
   at PnP.Core.Auth.Services.AuthenticationProviderFactory.Create(String name)
   at PnP.Core.Auth.Services.Builder.Configuration.AuthenticationProvidersOptionsConfigurator.Configure(PnPCoreOptions options)
   at Microsoft.Extensions.Options.OptionsFactory`1.Create(String name)
   at Microsoft.Azure.WebJobs.Hosting.WebJobsOptionsFactory`1.Create(String name) in C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Hosting\OptionsFormatter\WebJobsOptionsFactory.cs:line 36

**Reply by jansenbe** [reply]
@juliemturner : I'll have a look at this tomorrow.

**Reply by jansenbe** [reply]
@juliemturner : did spin up a quick test and below code works for me as v3 function.

`Startup.cs`:

```csharp
using Microsoft.Azure.Functions.Extensions.DependencyInjection;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using PnP.Core.Auth;
using PnP.Core.Auth.Services.Builder.Configuration;
using PnP.Core.Services.Builder.Configuration;
using System;
using System.Security.Cryptography.X509Certificates;

[assembly: FunctionsStartup(typeof(v3functionappforjulie.Startup))]

namespace v3functionappforjulie
{
    public class Startup : FunctionsStartup
    {
        public override void Configure(IFunctionsHostBuilder builder)
        {
            var config = builder.GetContext().Configuration;
            var azureFunctionSettings = new AzureFunctionSettings();
            config.Bind(azureFunctionSettings);

            builder.Services.AddPnPCore(options =>
            {
                // Disable telemetry because of mixed versions on AppInsights dependencies
                options.DisableTelemetry = true;

                // Configure an authentication provider with certificate (Required for app only)
                var authProvider = new X509CertificateAuthenticationProvider(azureFunctionSettings.ClientId,
                    azureFunctionSettings.TenantId,
                    LoadCertificate(azureFunctionSettings));

                // And set it as default
                options.DefaultAuthenticationProvider = authProvider;

                // Add a default configuration with the site configured in app settings
                options.Sites.Add("Default",
                       new PnPCoreSiteOptions
                       {
                           SiteUrl = azureFunctionSettings.SiteUrl,
                           AuthenticationProvider = authProvider
                       });
            });
        }

        private static X509Certificate2 LoadCertificate(AzureFunctionSettings azureFunctionSettings)
        {
            // Will only be populated correctly when running in the Azure Function host
            string certBase64Encoded = Environment.GetEnvironmentVariable("CertificateFromKeyVault");

            if (!string.IsNullOrEmpty(certBase64Encoded))
            {
                // Azure Function flow
                return new X509Certificate2(Convert.FromBase64String(certBase64Encoded),
                                            "",
                                            X509KeyStorageFlags.Exportable |
                                            X509KeyStorageFlags.MachineKeySet |
                                            X509KeyStorageFlags.EphemeralKeySet);
            }
            else
            {
                // Local flow
                var store = new X509Store(StoreName.My, StoreLocation.CurrentUser);
                store.Open(OpenFlags.ReadOnly | OpenFlags.OpenExistingOnly);
                var certificateCollection = store.Certificates.Find(X509FindType.FindByThumbprint, azureFunctionSettings.CertificateThumbprint, false);
                store.Close();

                return certificateCollection[0];
            }
        }
    }
}
```` 

`Function1.cs`:

```csharp
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Azure.WebJobs;
using Microsoft.Azure.WebJobs.Extensions.Http;
using Microsoft.Extensions.Logging;
using PnP.Core.Model.SharePoint;
using PnP.Core.Services;
using System.Linq;
using System.Threading.Tasks;

namespace v3functionappforjulie
{
    public class Function1
    {

        private readonly ILogger logger;
        private readonly IPnPContextFactory contextFactory;

        public Function1(IPnPContextFactory pnpContextFactory, ILoggerFactory loggerFactory)
        {
            logger = loggerFactory.CreateLogger<Function1>();
            contextFactory = pnpContextFactory;
        }

        [FunctionName("Function1")]
        public async Task<IActionResult> Run(
            [HttpTrigger(AuthorizationLevel.Function, "get", "post", Route = null)] HttpRequest req,
            ILogger log)
        {

            using (var pnpContext = await contextFactory.CreateAsync(new System.Uri("https://contoso.sharepoint.com/sites/mysitetouse")))
            {
                log.LogInformation("Getting all the shared documents on site.");

                IList sharedDocuments = await pnpContext.Web.Lists.GetByTitleAsync("Documents", l => l.RootFolder);
                var sharedDocumentsFolder = await sharedDocuments.RootFolder.GetAsync(f => f.Files);

                var documents = (from d in sharedDocumentsFolder.Files.AsEnumerable()
                                 select new
                                 {
                                     d.Name,
                                     d.TimeLastModified,
                                     d.UniqueId
                                 }).ToList();

                return new OkObjectResult(new { documents });
            }
        }
    }
}
```

**Reply by jansenbe** [reply]
I could also reproduce your issue, something I still have to look into. Seems related to differences in how dependency injection works between the old .NET Core and .NET 5/6

**Reply by jansenbe** [reply]
@juliemturner : if things work for you then feel free mark this thread as answered.

**Reply by juliemturner** [reply]
Nothing is working for me no, I have the issue I last posted and have been waiting for guidance.  You said you could reproduce it so I assumed that means it wasn't just me.

**Reply by juliemturner** [reply]
Ok, wait I see you posted something above that's different from what I did.  I didn't notice that I only got the email notification that said you could replicate the issue, so I went into waiting mode and worked on something else.  Ok, I'll try that out.

**Reply by juliemturner** [reply]
Ok, wait I see you posted something above that's different from what I did.  I didn't notice that I only got the email notification that said you could replicate the issue, so I went into waiting mode and worked on something else.  

Ok, the code you posted above does work so I guess I'm all good.  Thanks for the help sorry I didn't notice it sooner.
